sources/api/corndog/src/main.rs

```diff
@@ -104,23 +109,83 @@ where
-/// path, and the value its contents.
-fn set_sysctls<K>(sysctls: HashMap<K, String>)
+/// Validates sysctl key/value pairs for systemd-sysctl compatibility
+fn validate_sysctl_entry<K>(key: K, value: &str) -> Result<()>
+where
+    K: AsRef<str>,
+{
+    let key = key.as_ref();
+
+    // Basic validation of key format
+    if !key.chars().all(|c| c.is_ascii_alphanumeric() || c == '.' || c == '_' || c == '-' || c == '*') {
+        return error::InvalidSysctlKeySnafu { key: key.to_string() }.fail();
+    }
+
+    // Ensure value doesn't contain newlines which could break the config file format
+    if value.contains('\n') {
+        return error::InvalidSysctlValueSnafu {
+            key: key.to_string(),
+            value: value.to_string()
+        }.fail();
+    }
+
+    Ok(())
+}
+
+/// Restarts systemd-sysctl service
+fn restart_sysctl() -> Result<()> {
+    let status = Command::new("systemctl")
+        .args(["restart", "systemd-sysctl"])
+        .status()
+        .context(error::SystemctlSnafu)?;
+
+    if !status.success() {
+        return error::SystemctlFailedSnafu.fail();
+    }
+
+    Ok(())
+}
+
+/// Applies the requested sysctls to the system using systemd-sysctl.
+fn set_sysctls<K>(sysctls: HashMap<K, String>) -> Result<()>
 where
     K: AsRef<str>,
 {
+    // TODO: make the priority configurable?
+    const SYSCTL_CONFIG: &str = "/etc/sysctl.d/95-corndog.conf";
+
+    // Validate all entries first
+    for (key, value) in &sysctls {
+        validate_sysctl_entry(key, value)?;
+    }
+
+    // Create config string
+    let mut config = String::new();
     for (key, value) in sysctls {
-        let key = key.as_ref();
-        let path = sysctl_path(key);
-        if let Err(e) = fs::write(path, value) {
-            // We don't fail because sysctl keys can vary between kernel versions and depend on
-            // loaded modules.  It wouldn't be possible to deploy settings to a mixed-kernel fleet
-            // if newer sysctl values failed on your older kernels, for example, and we believe
-            // it's too cumbersome to have to specify in settings which keys are allowed to fail.
-            error!("Failed to write sysctl value '{}': {}", key, e);
-        }
+        config.push_str(&format!("{}={}\n", key.as_ref(), value));
     }
+
+    // Append to file, creating it if it doesn't exist
+    let mut file = fs::OpenOptions::new()
+        .create(true)
+        .append(true)
+        .open(SYSCTL_CONFIG)
+        .context(error::OpenSysctlConfigSnafu)?;
+
+    file.write_all(config.as_bytes())
+        .context(error::WriteSysctlConfigSnafu)?;
+
+    // Restart systemd-sysctl to apply changes
+    let status = Command::new("systemctl")
+        .args(["restart", "systemd-sysctl"])
+        .status()
+        .context(error::SystemctlSnafu)?;
+
+    if !status.success() {
+        return error::SystemctlFailedSnafu.fail();
+    }
```

Comment:
When hitting this case, it would be also good to print the errors from the output. So I would get the output of the command and check the status and the stderr.

```
    let output = Command::new(SYSTEMD_SYSCTL_BIN)
        .arg(SYSCTL_CONFIG_PATH)
        .output()
        .context(ExecutionSnafu{command: format!("{} {}", SYSTEMD_SYSCTL_BIN, SYSCTL_CONFIG_PATH)})?;
    if !output.status.success() {
        warn!(
            "Failed to apply sysctl config: {}",
            String::from_utf8_lossy(&output.stderr) # account for invalid utf8 encoding
        );
    }
```